src/main/java/org/jenkinsci/main/modules/instance_identity/InstanceIdentity.java

```diff
@@ -53,19 +54,32 @@ public InstanceIdentity(File keyFile) throws IOException {
             keys = read(new FileReader(oldKeyFile), gen);
             write(keys, keyFile);
             Util.deleteFile(oldKeyFile);
-        } else if (keyFile.exists()) {
-            byte[] enc = FileUtils.readFileToByteArray(keyFile);
+        } else {
+            Reader reader = getFileReader(keyFile);
+            if (reader!=null) {
+                keys = read(reader, gen);
+            } else {
+                gen.initialize(2048, new SecureRandom()); // going beyond 2048 requires crypto extension
+                keys = gen.generateKeyPair();
+                write(keys, keyFile);
+            }
+        }
+    }
```

Comment:
Am I right in saying that the `read` is only used from two places inside this class and in both cases the `Reader`s being passed to it are coming from `File` instances (`oldKeyFile` and `keyFile `) ?

If so, then maybe it would not be easier/cleaner to just change the `read` function to take a `File` and then, from inside `read`, check if the file exists, read it's contents and return the `keys` if it can, or fail as appropriate (`returning `null` or whatever). If reading from `keyFile` and the `read` returns `null` then you can go through the `gen.initialize` path.